src/main/java/com/google/devtools/build/lib/bazel/bzlmod/GsonTypeAdapterUtil.java

```diff
@@ -283,6 +301,68 @@ public Optional<T> read(JsonReader jsonReader) throws IOException {
     }
   }
 
+  /**
+   * Converts Guava tables into a JSON array of 3-tuples (one per cell). Each 3-tuple is a JSON
+   * array itself (rowKey, columnKey, value).
+   */
+  public static final TypeAdapterFactory IMMUTABLE_TABLE =
+      new TypeAdapterFactory() {
+        @Nullable
+        @Override
+        @SuppressWarnings("unchecked")
+        public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> typeToken) {
+          if (typeToken.getRawType() != ImmutableTable.class) {
+            return null;
+          }
+          Type type = typeToken.getType();
+          if (!(type instanceof ParameterizedType)) {
+            return null;
+          }
+          Type[] typeArgs = ((ParameterizedType) typeToken.getType()).getActualTypeArguments();
+          if (typeArgs.length != 3) {
+            return null;
+          }
+          var typeArgAdapters = Arrays.stream(typeArgs)
```

Comment:
Instead of a list of `TypeAdapter` objects, where you have to call `typeArgAdapters.get(0)`, why not have separate variables for each of `rowTypeAdapter`, `columnTypeAdapter`, and `valueTypeAdapter`?

Reply:
you have to get the type arguments out of an array anyway. I thought this was better than writing `rowTypeAdapter = blahblahblah(typeArgs[0]); columnTypeAdapter = blahblahblah(typeArgs[1]); ...` since I could use `stream.map`.

Reply:
I don't think that the benefit of using `stream.map` outweighs the confusion of having to find the different adapters by an integer index multiple times. There's nothing currently ensuring that the same adapter is used in both the read and write methods except for author diligence. While that is technically also true for `rowTypeAdapter` etc, it's much more obvious there when a mistake is being made.

Reply:
that's fair -- I updated it to use three separate variables